src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,486 @@
+    const SolverIterate solver_iterate, const SolverFree solver_free) {
+  // Check for valid stopping condition:
+  if (UNLIKELY(condition != StoppingCondition::AbsoluteAndRelative and
+               condition != StoppingCondition::Absolute)) {
+    ERROR(
+        "Invalid stopping condition. Has to be either AbsoluteAndRelative"
+        "or Absolute.");
+  }
+
+  // Supply gsl_root with the initial guess:
+  gsl_vector* const gsl_root = gsl_vector_alloc(Dim);
+  gsl_vector_set_with_std_array(gsl_root, initial_guess);
+  auto* const solver = solver_alloc(solver_type, Dim);
+  solver_set(solver, &f, gsl_root);
+
+  // Take iterations:
+  int status;
+  size_t iteration_number = 0;
+  do {
+    if (UNLIKELY(verbosity == Verbosity::Debug)) {
+      print_state<Dim>(iteration_number, solver, iteration_number == 0);
+    }
+    iteration_number++;
+    status = solver_iterate(solver);
+    // Check if solver is stuck
+    if (UNLIKELY(status == GSL_ENOPROG)) {
+      if (UNLIKELY(verbosity == Verbosity::Debug)) {
+        Parallel::printf(
+            "The iteration is not making any progress, preventing the "
+            "algorithm from continuing.");
+      }
+      break;
+    }
+    if (condition == StoppingCondition::AbsoluteAndRelative) {
+      status = gsl_multiroot_test_delta(solver->dx, solver->x,
+                                        absolute_tolerance, relative_tolerance);
+    } else {  // condition is StoppingCondition::Absolute
+      // NOTE: Sometimes hybridsj works only with the test_residual condition
+      status = gsl_multiroot_test_residual(solver->f, absolute_tolerance);
+    }
+  } while (status == GSL_CONTINUE and iteration_number < maximum_iterations);
+  if (UNLIKELY(verbosity == Verbosity::Verbose or
+               verbosity == Verbosity::Debug)) {
+    print_state<Dim>(iteration_number, solver, verbosity == Verbosity::Verbose);
+  }
+  bool success = (status == GSL_SUCCESS);
+  if (UNLIKELY(verbosity != Verbosity::Silent)) {
+    Parallel::printf("\n");
+    if (not success) {
+      const std::string ascii_divider = std::string(70, '#');
+      const std::string failure_message =
+          ascii_divider + "\n\t\t\tWARNING: Root Finding FAILED\n" +
+          ascii_divider;
+      Parallel::printf("%s\n", failure_message);
+    } else {
+      Parallel::printf("Root finder converged.\n");
+    }
+  }
+  // If maximum_absolute_tolerance is given, return success = true
+  // as long as maximum_absolute_tolerance is achieved even if the
+  // root finder doesn't converge.
+  bool success_with_tolerance = true;
+  bool failed_root_is_forgiven = false;
+  if (not success and maximum_absolute_tolerance > 0) {
+    for (size_t i = 0; i < Dim; ++i) {
+      if (fabs(gsl_vector_get(solver->f, i)) > maximum_absolute_tolerance) {
+        success_with_tolerance = false;
+      }
+    }
+    failed_root_is_forgiven = success_with_tolerance;
+  }
+  if (success_with_tolerance) {
+    success = true;
+  }
```

Comment:
If I am reading this logic correctly, this will always report success if `maximum_absolute_tolerance == 0.` (the default).

Add tests that the code correctly reports failure when it fails.